src/main/java/com/exactpro/th2/mstore/ErrorCollector.java

```diff
@@ -0,0 +1,185 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.exactpro.th2.mstore;
+
+import com.exactpro.th2.common.event.Event;
+import com.exactpro.th2.common.event.Event.Status;
+import com.exactpro.th2.common.event.IBodyData;
+import com.exactpro.th2.common.grpc.EventBatch;
+import com.exactpro.th2.common.grpc.EventID;
+import com.exactpro.th2.common.schema.message.MessageRouter;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonFormat;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.datatype.jsr310.ser.InstantSerializer;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static java.util.Objects.requireNonNull;
+
+@SuppressWarnings("unused")
+public class ErrorCollector implements AutoCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ErrorCollector.class);
+    private static final ThreadFactory THREAD_FACTORY = new ThreadFactoryBuilder().setNameFormat("error-collecor-%d").build();
+    private final ScheduledFuture<?> drainFuture;
+    private final MessageRouter<EventBatch> eventRouter;
+    private final EventID rootEvent;
+    private final Lock lock = new ReentrantLock();
+    private Map<String, ErrorMetadata> errors = new HashMap<>();
+
+    public ErrorCollector(@NotNull ScheduledExecutorService executor,
+                          @NotNull MessageRouter<EventBatch> eventRouter,
+                          EventID rootEvent,
+                          long period,
+                          TimeUnit unit) {
+        this.eventRouter = requireNonNull(eventRouter, "Event router can't be null");
+        this.rootEvent = requireNonNull(rootEvent, "Root event can't be null");
+        this.drainFuture = requireNonNull(executor, "Executor can't be null")
+                .scheduleAtFixedRate(this::drain, period, period, unit);
```

Comment:
I believe that the task submission should be done outside. Instead, the ErrorCollector class should expose `drain` method that can be called to drain all the errors. This also will help with the testability fo the class.

However, it is up to you to decide. I do not insist on that

Reply:
I changed drain method visibility, but stay subscription inside. We can pass mock instead of executor and call drain method from code in tests

Reply:
Thank you for advice, but I would like to stay my approach

Reply:
Okay, as I said it is up to you